File: node/src/providers/protocol-adapter.ts
import type {
  Message,
  ProviderReplay,
  ProviderRunState,
  ProviderUsage,
  StreamEvent,
} from "../types.js"
import type { CanonicalAdapterInput } from "./content-normalization.js"
import type {
  GenerationProtocol,
  ProtocolRuntimeCapabilities,
} from "./protocol-capabilities.js"

export {
  GEMINI_PROTOCOL_CAPABILITIES,
  OLLAMA_PROTOCOL_CAPABILITIES,
} from "./protocol-capabilities.js"

export type CanonicalStopReason =
  | "end_turn"
  | "tool_use"
  | "max_tokens"
  | "stop_sequence"
  | "content_filter"
  | "other"

export interface AdapterOutput {
  events: StreamEvent[]
  replay?: ProviderReplay
  runStatePatch?: Partial<ProviderRunState>
}

export interface AdapterDecodeInput {
  input: CanonicalAdapterInput
}

export interface AdapterStreamInput {
  input: CanonicalAdapterInput
}

export interface ProtocolAdapter<
  TRequest,
  TCompleteResponse,
  TStreamChunk,
  TStreamState,
  TStreamFinal = undefined,
> {
  readonly protocol: GenerationProtocol
  readonly protocolCapabilities: ProtocolRuntimeCapabilities

  buildRequest(input: CanonicalAdapterInput): TRequest
  decodeComplete(raw: TCompleteResponse, input: AdapterDecodeInput): {
    message: Message
    replay?: ProviderReplay
  }

  createStreamState(input: AdapterStreamInput): TStreamState
  pushStreamChunk(chunk: TStreamChunk, state: TStreamState): AdapterOutput
  finishStream(
    state: TStreamState,
    final: TStreamFinal,
  ): AdapterOutput | Promise<AdapterOutput>

  normalizeUsage(raw: unknown): ProviderUsage | undefined
  normalizeStopReason(raw: string | undefined): CanonicalStopReason | undefined
}

export class ProtocolResponseError extends Error {
  readonly protocol: GenerationProtocol

  constructor(protocol: GenerationProtocol, message: string) {
    super(`${protocol} protocol response error: ${message}`)
    this.name = "ProtocolResponseError"
    this.protocol = protocol
  }
}

File: python/deepstrike/providers/protocol_adapter.py
"""Typed provider-protocol lifecycle contracts.

Adapters convert a validated canonical input to one wire protocol. Transport ownership
(SDK clients, retries, circuit breakers, and durable replay/state stores) stays with
the provider runtime.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Generic, Protocol, TypeVar

from deepstrike._kernel import Message
from deepstrike.providers.stream import StreamEvent
from deepstrike.types.content import CanonicalAdapterInput

RequestT = TypeVar("RequestT")
CompleteT = TypeVar("CompleteT")
ChunkT = TypeVar("ChunkT")
StateT = TypeVar("StateT")
FinalT = TypeVar("FinalT")


@dataclass(frozen=True)
class AdapterOutput:
    events: list[StreamEvent] = field(default_factory=list)
    replay: dict[str, Any] | None = None
    run_state_patch: dict[str, Any] | None = None


@dataclass(frozen=True)
class AdapterDecodeInput:
    input: CanonicalAdapterInput


@dataclass(frozen=True)
class AdapterStreamInput:
    input: CanonicalAdapterInput


class ProtocolResponseError(ValueError):
    def __init__(self, protocol: str, message: str):
        self.protocol = protocol
        super().__init__(f"{protocol} protocol response error: {message}")


class ProtocolAdapter(Protocol, Generic[RequestT, CompleteT, ChunkT, StateT, FinalT]):
    protocol: str

    def build_request(self, input: CanonicalAdapterInput) -> RequestT: ...
    def decode_complete(self, raw: CompleteT, input: CanonicalAdapterInput) -> Message: ...
    def create_stream_state(self, input: CanonicalAdapterInput) -> StateT: ...
    def push_stream_chunk(self, chunk: ChunkT, state: StateT) -> AdapterOutput: ...
    def finish_stream(self, state: StateT, final: FinalT) -> AdapterOutput: ...
